{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import islice, combinations, product\n",
    "from collections import defaultdict\n",
    "from goatools import obo_parser\n",
    "from functools import partial\n",
    "\n",
    "import os\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "sns.set()\n",
    "pd.set_option(\"display.max_columns\", 50)\n",
    "\n",
    "DATA_DIRECTORY = \"/media/clusterduck123/joe/data\"\n",
    "RAW_DATA_DIRECTORY = f\"{DATA_DIRECTORY}/raw_data\"\n",
    "YEAST_DIRECTORY = f\"{DATA_DIRECTORY}/processed_data/yeast\"\n",
    "NETWORK_DIRECTORY = f\"{YEAST_DIRECTORY}/networks\"\n",
    "MATRIX_DIRECTORY  = f\"{YEAST_DIRECTORY}/distance-matrices\"\n",
    "ANNOTATION_DIRECTORY  = f\"{YEAST_DIRECTORY}/annotations\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "aspect = 'BP'\n",
    "correction = 'BH'\n",
    "\n",
    "alpha = 0.05\n",
    "lb_GO = 5\n",
    "ub_GO = 500\n",
    "min_lvl = 0\n",
    "max_lvl = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/clusterduck123/joe/data/raw_data/go-basic.obo: fmt(1.2) rel(2019-10-07) 47,285 GO Terms\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'annotation_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-bfe620bc4fa8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mgene_population\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mPPI\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnodes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m GO_population = {go_id for go_id in set(annotation_df.GO_ID) \n\u001b[0m\u001b[1;32m      9\u001b[0m                            if (lb_GO <= len(annotation_df[annotation_df.GO_ID == go_id]) <= ub_GO and\n\u001b[1;32m     10\u001b[0m                                min_lvl <= go_dag[go_id].level <= max_lvl)}\n",
      "\u001b[0;31mNameError\u001b[0m: name 'annotation_df' is not defined"
     ]
    }
   ],
   "source": [
    "PPI = nx.read_edgelist(f\"{NETWORK_DIRECTORY}/official_PPI_BioGRID.txt\")\n",
    "\n",
    "#annotation_df = pd.read_csv(f\"{ANNOTATION_DIRECTORY}/GO_{aspect}_BioGRID-SGD.csv\")\n",
    "\n",
    "go_dag = obo_parser.GODag(f\"{RAW_DATA_DIRECTORY}/go-basic.obo\")\n",
    "\n",
    "gene_population = set(PPI.nodes())\n",
    "GO_population = {go_id for go_id in set(annotation_df.GO_ID) \n",
    "                           if (lb_GO <= len(annotation_df[annotation_df.GO_ID == go_id]) <= ub_GO and\n",
    "                               min_lvl <= go_dag[go_id].level <= max_lvl)}\n",
    "print(len(GO_population))\n",
    "\n",
    "annotation_df = annotation_df[annotation_df.GO_ID.isin(GO_population)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['regulation of DNA recombination',\n",
       " 'regulation of mitotic recombination',\n",
       " 'regulation of cyclin-dependent protein serine/threonine kinase activity',\n",
       " 'negative regulation of transcription by RNA polymerase II',\n",
       " 'negative regulation of transposition, DNA-mediated',\n",
       " 'positive regulation of transposition, DNA-mediated',\n",
       " 'regulation of alternative mRNA splicing, via spliceosome',\n",
       " 'regulation of transcription by galactose',\n",
       " 'positive regulation of transcription by galactose',\n",
       " 'regulation of histone H3-K36 methylation',\n",
       " 'negative regulation of histone H3-K36 methylation',\n",
       " 'positive regulation of histone H3-K36 methylation',\n",
       " 'carbon catabolite regulation of transcription from RNA polymerase II promoter',\n",
       " 'regulation of transcription from RNA polymerase II promoter by glucose',\n",
       " 'regulation of transcription from RNA polymerase II promoter by galactose',\n",
       " 'positive regulation of transcription from RNA polymerase II promoter by glucose',\n",
       " 'positive regulation of transcription from RNA polymerase II promoter by galactose',\n",
       " 'regulation of glutamine family amino acid metabolic process',\n",
       " 'regulation of arginine metabolic process',\n",
       " 'regulation of mitochondrial RNA catabolic process',\n",
       " 'negative regulation of mitochondrial RNA catabolic process',\n",
       " 'positive regulation of mitochondrial RNA catabolic process',\n",
       " 'transcription regulatory region sequence-specific DNA binding',\n",
       " 'RNA polymerase II regulatory region sequence-specific DNA binding',\n",
       " 'bacterial-type RNA polymerase regulatory region sequence-specific DNA binding',\n",
       " 'polymerase III regulatory region sequence-specific DNA binding',\n",
       " 'RNA polymerase II regulatory region DNA binding',\n",
       " 'RNA polymerase I regulatory region DNA binding',\n",
       " 'RNA polymerase III regulatory region DNA binding',\n",
       " 'bacterial-type RNA polymerase regulatory region DNA binding',\n",
       " 'plastid RNA polymerase regulatory region DNA binding',\n",
       " 'transcription regulatory region RNA binding',\n",
       " 'negative regulation of exit from mitosis',\n",
       " 'intronic transcription regulatory region sequence-specific DNA binding',\n",
       " 'RNA polymerase II intronic transcription regulatory region sequence-specific DNA binding',\n",
       " 'RNA polymerase I regulatory region sequence-specific DNA binding',\n",
       " 'regulation of transcriptional start site selection at RNA polymerase II promoter',\n",
       " 'regulation of mating-type specific transcription from RNA polymerase II promoter',\n",
       " 'positive regulation of mating-type specific transcription from RNA polymerase II promoter',\n",
       " 'negative regulation of mating-type specific transcription from RNA polymerase II promoter',\n",
       " 'negative regulation of chromatin silencing involved in replicative cell aging',\n",
       " 'regulation of chromatin assembly or disassembly',\n",
       " 'regulation of type IIa hypersensitivity',\n",
       " 'negative regulation of type IIa hypersensitivity',\n",
       " 'positive regulation of type IIa hypersensitivity',\n",
       " 'regulation of type IIb hypersensitivity',\n",
       " 'negative regulation of type IIb hypersensitivity',\n",
       " 'positive regulation of type IIb hypersensitivity',\n",
       " 'regulation of type III hypersensitivity',\n",
       " 'negative regulation of type III hypersensitivity',\n",
       " 'positive regulation of type III hypersensitivity',\n",
       " 'regulation of type IV hypersensitivity',\n",
       " 'negative regulation of type IV hypersensitivity',\n",
       " 'positive regulation of type IV hypersensitivity',\n",
       " 'regulation of type I hypersensitivity',\n",
       " 'negative regulation of type I hypersensitivity',\n",
       " 'positive regulation of type I hypersensitivity',\n",
       " 'negative regulation of antibody-dependent cellular cytotoxicity',\n",
       " 'positive regulation of antibody-dependent cellular cytotoxicity',\n",
       " 'negative regulation of complement activation, lectin pathway',\n",
       " 'positive regulation of complement activation, lectin pathway',\n",
       " 'negative regulation of cytolysis by symbiont of host cells',\n",
       " 'positive regulation of cytolysis by symbiont of host cells',\n",
       " 'negative regulation of T cell mediated cytotoxicity',\n",
       " 'positive regulation of T cell mediated cytotoxicity',\n",
       " 'regulation of B-1 B cell differentiation',\n",
       " 'negative regulation of B-1 B cell differentiation',\n",
       " 'positive regulation of B-1 B cell differentiation',\n",
       " 'negative regulation of exocyst assembly',\n",
       " 'positive regulation of exocyst assembly',\n",
       " 'regulation of protein phosphorylation',\n",
       " 'negative regulation of protein phosphorylation',\n",
       " 'positive regulation of protein phosphorylation',\n",
       " 'regulation of endothelial cell proliferation',\n",
       " 'negative regulation of endothelial cell proliferation',\n",
       " 'positive regulation of endothelial cell proliferation',\n",
       " 'negative regulation of cell-matrix adhesion',\n",
       " 'positive regulation of cell-matrix adhesion',\n",
       " 'negative regulation of cytokine-mediated signaling pathway',\n",
       " 'positive regulation of cytokine-mediated signaling pathway',\n",
       " 'regulation of activation of membrane attack complex',\n",
       " 'positive regulation of activation of membrane attack complex',\n",
       " 'negative regulation of activation of membrane attack complex',\n",
       " 'negative regulation of the force of heart contraction involved in baroreceptor response to increased systemic arterial blood pressure',\n",
       " 'positive regulation of the force of heart contraction involved in baroreceptor response to decreased systemic arterial blood pressure',\n",
       " 'regulation of systemic arterial blood pressure by circulatory renin-angiotensin',\n",
       " 'regulation of systemic arterial blood pressure by norepinephrine-epinephrine',\n",
       " 'norepinephrine-epinephrine vasoconstriction involved in regulation of systemic arterial blood pressure',\n",
       " 'positive regulation of the force of heart contraction by epinephrine-norepinephrine',\n",
       " 'angiotensin-mediated vasoconstriction involved in regulation of systemic arterial blood pressure',\n",
       " 'secretion of vasopressin involved in fast regulation of systemic arterial blood pressure',\n",
       " 'regulation of blood volume by renin-angiotensin',\n",
       " 'regulation of renal output by angiotensin',\n",
       " 'norepinephrine-epinephrine-mediated vasodilation involved in regulation of systemic arterial blood pressure',\n",
       " 'regulation of sodium ion transport',\n",
       " 'angiotensin-mediated vasodilation involved in regulation of systemic arterial blood pressure',\n",
       " 'regulation of blood vessel diameter by renin-angiotensin',\n",
       " 'negative regulation of L-glutamate import across plasma membrane',\n",
       " 'positive regulation of L-glutamate import across plasma membrane',\n",
       " 'positive regulation of neuroblast proliferation',\n",
       " 'positive regulation of mesenchymal cell proliferation',\n",
       " 'negative regulation of receptor internalization',\n",
       " 'positive regulation of receptor internalization',\n",
       " 'negative regulation of thyroid hormone mediated signaling pathway',\n",
       " 'positive regulation of thyroid hormone mediated signaling pathway',\n",
       " 'positive regulation of antimicrobial peptide production',\n",
       " 'positive regulation of defense response to virus by host',\n",
       " 'negative regulation of kinin cascade',\n",
       " 'positive regulation of kinin cascade',\n",
       " 'CD4-positive, CD25-positive, alpha-beta regulatory T cell differentiation involved in immune response',\n",
       " 'CD8-positive, alpha-beta regulatory T cell differentiation',\n",
       " 'CD4-positive, CD25-positive, alpha-beta regulatory T cell differentiation',\n",
       " 'CD4-positive, CD25-positive, alpha-beta regulatory T cell lineage commitment',\n",
       " 'regulation of tissue kallikrein-kinin cascade',\n",
       " 'regulation of plasma kallikrein-kinin cascade',\n",
       " 'regulation of heart contraction involved in acute-phase response',\n",
       " 'negative regulation of tissue kallikrein-kinin cascade',\n",
       " 'positive regulation of tissue kallikrein-kinin cascade',\n",
       " 'negative regulation of plasma kallikrein-kinin cascade',\n",
       " 'positive regulation of plasma kallikrein-kinin cascade',\n",
       " 'negative regulation of antigen processing and presentation of peptide or polysaccharide antigen via MHC class II',\n",
       " 'positive regulation of antigen processing and presentation of peptide or polysaccharide antigen via MHC class II',\n",
       " 'negative regulation of antigen processing and presentation of peptide antigen',\n",
       " 'positive regulation of antigen processing and presentation of peptide antigen',\n",
       " 'regulation of antigen processing and presentation of peptide antigen via MHC class II',\n",
       " 'negative regulation of antigen processing and presentation of peptide antigen via MHC class II',\n",
       " 'positive regulation of antigen processing and presentation of peptide antigen via MHC class II',\n",
       " 'regulation of antigen processing and presentation of peptide antigen via MHC class I',\n",
       " 'negative regulation of antigen processing and presentation of peptide antigen via MHC class I',\n",
       " 'positive regulation of antigen processing and presentation of peptide antigen via MHC class I',\n",
       " 'negative regulation of antigen processing and presentation via MHC class Ib',\n",
       " 'positive regulation of antigen processing and presentation via MHC class Ib',\n",
       " 'regulation of antigen processing and presentation of peptide antigen via MHC class Ib',\n",
       " 'negative regulation of antigen processing and presentation of peptide antigen via MHC class Ib',\n",
       " 'positive regulation of antigen processing and presentation of peptide antigen via MHC class Ib',\n",
       " 'regulation of antigen processing and presentation of lipid antigen via MHC class Ib',\n",
       " 'negative regulation of antigen processing and presentation of lipid antigen via MHC class Ib',\n",
       " 'positive regulation of antigen processing and presentation of lipid antigen via MHC class Ib',\n",
       " 'regulation of antigen processing and presentation of polysaccharide antigen via MHC class II',\n",
       " 'negative regulation of antigen processing and presentation of polysaccharide antigen via MHC class II',\n",
       " 'positive regulation of antigen processing and presentation of polysaccharide antigen via MHC class II',\n",
       " 'negative regulation of dendritic cell antigen processing and presentation',\n",
       " 'positive regulation of dendritic cell antigen processing and presentation',\n",
       " 'regulation of myeloid dendritic cell antigen processing and presentation',\n",
       " 'negative regulation of myeloid dendritic cell antigen processing and presentation',\n",
       " 'positive regulation of myeloid dendritic cell antigen processing and presentation',\n",
       " 'regulation of plasmacytoid dendritic cell antigen processing and presentation',\n",
       " 'negative regulation of plasmacytoid dendritic cell antigen processing and presentation',\n",
       " 'positive regulation of plasmacytoid dendritic cell antigen processing and presentation',\n",
       " 'negative regulation of monocyte antigen processing and presentation',\n",
       " 'positive regulation of monocyte antigen processing and presentation',\n",
       " 'negative regulation of macrophage antigen processing and presentation',\n",
       " 'positive regulation of macrophage antigen processing and presentation',\n",
       " 'negative regulation of non-professional antigen presenting cell antigen processing and presentation',\n",
       " 'positive regulation of non-professional antigen presenting cell antigen processing and presentation',\n",
       " 'negative regulation of B cell antigen processing and presentation',\n",
       " 'positive regulation of B cell antigen processing and presentation',\n",
       " 'negative regulation of T cell antigen processing and presentation',\n",
       " 'positive regulation of T cell antigen processing and presentation',\n",
       " 'regulation of proteolysis associated with antigen processing and presentation',\n",
       " 'negative regulation of proteolysis associated with antigen processing and presentation',\n",
       " 'positive regulation of proteolysis associated with antigen processing and presentation',\n",
       " 'negative regulation of granuloma formation',\n",
       " 'positive regulation of granuloma formation',\n",
       " 'regulation of immunoglobulin production',\n",
       " 'negative regulation of immunoglobulin production',\n",
       " 'positive regulation of immunoglobulin production',\n",
       " 'regulation of immunoglobulin biosynthetic process',\n",
       " 'negative regulation of immunoglobulin biosynthetic process',\n",
       " 'positive regulation of immunoglobulin biosynthetic process',\n",
       " 'negative regulation of central tolerance induction',\n",
       " 'positive regulation of central tolerance induction',\n",
       " 'negative regulation of tolerance induction to self antigen',\n",
       " 'positive regulation of tolerance induction to self antigen',\n",
       " 'negative regulation of tolerance induction dependent upon immune response',\n",
       " 'positive regulation of tolerance induction dependent upon immune response',\n",
       " 'regulation of tolerance induction to nonself antigen',\n",
       " 'negative regulation of tolerance induction to nonself antigen',\n",
       " 'positive regulation of tolerance induction to nonself antigen',\n",
       " 'regulation of peripheral tolerance induction',\n",
       " 'negative regulation of peripheral tolerance induction',\n",
       " 'positive regulation of peripheral tolerance induction',\n",
       " 'negative regulation of B cell tolerance induction',\n",
       " 'positive regulation of B cell tolerance induction',\n",
       " 'negative regulation of T cell tolerance induction',\n",
       " 'positive regulation of T cell tolerance induction',\n",
       " 'regulation of T cell anergy',\n",
       " 'negative regulation of T cell anergy',\n",
       " 'positive regulation of T cell anergy',\n",
       " 'regulation of B cell anergy',\n",
       " 'negative regulation of B cell anergy',\n",
       " 'positive regulation of B cell anergy',\n",
       " 'regulation of acute inflammatory response',\n",
       " 'negative regulation of acute inflammatory response',\n",
       " 'positive regulation of acute inflammatory response',\n",
       " 'regulation of chronic inflammatory response',\n",
       " 'negative regulation of chronic inflammatory response',\n",
       " 'positive regulation of chronic inflammatory response',\n",
       " 'negative regulation of leukocyte chemotaxis',\n",
       " 'positive regulation of leukocyte chemotaxis',\n",
       " 'negative regulation of cellular extravasation',\n",
       " 'positive regulation of cellular extravasation',\n",
       " 'negative regulation of production of molecular mediator of immune response',\n",
       " 'positive regulation of production of molecular mediator of immune response',\n",
       " 'negative regulation of leukocyte mediated immunity',\n",
       " 'positive regulation of leukocyte mediated immunity',\n",
       " 'regulation of lymphocyte mediated immunity',\n",
       " 'negative regulation of lymphocyte mediated immunity',\n",
       " 'positive regulation of lymphocyte mediated immunity',\n",
       " 'regulation of T cell mediated immunity',\n",
       " 'negative regulation of T cell mediated immunity',\n",
       " 'positive regulation of T cell mediated immunity',\n",
       " 'regulation of B cell mediated immunity',\n",
       " 'negative regulation of B cell mediated immunity',\n",
       " 'positive regulation of B cell mediated immunity',\n",
       " 'negative regulation of natural killer cell mediated immunity',\n",
       " 'positive regulation of natural killer cell mediated immunity',\n",
       " 'negative regulation of cytokine production involved in immune response',\n",
       " 'positive regulation of cytokine production involved in immune response',\n",
       " 'regulation of B cell cytokine production',\n",
       " 'negative regulation of B cell cytokine production',\n",
       " 'positive regulation of B cell cytokine production',\n",
       " 'regulation of T cell cytokine production',\n",
       " 'negative regulation of T cell cytokine production',\n",
       " 'positive regulation of T cell cytokine production',\n",
       " 'regulation of natural killer cell cytokine production',\n",
       " 'negative regulation of natural killer cell cytokine production',\n",
       " 'positive regulation of natural killer cell cytokine production',\n",
       " 'regulation of dendritic cell cytokine production',\n",
       " 'negative regulation of dendritic cell cytokine production',\n",
       " 'positive regulation of dendritic cell cytokine production',\n",
       " 'regulation of myeloid dendritic cell cytokine production',\n",
       " 'negative regulation of myeloid dendritic cell cytokine production',\n",
       " 'positive regulation of myeloid dendritic cell cytokine production',\n",
       " 'regulation of plasmacytoid dendritic cell cytokine production',\n",
       " 'negative regulation of plasmacytoid dendritic cell cytokine production',\n",
       " 'positive regulation of plasmacytoid dendritic cell cytokine production',\n",
       " 'regulation of cytokine secretion involved in immune response',\n",
       " 'negative regulation of cytokine secretion involved in immune response',\n",
       " 'positive regulation of cytokine secretion involved in immune response',\n",
       " 'regulation of cytokine biosynthetic process involved in immune response',\n",
       " 'negative regulation of cytokine biosynthetic process involved in immune response',\n",
       " 'positive regulation of cytokine biosynthetic process involved in immune response',\n",
       " 'regulation of myeloid leukocyte differentiation',\n",
       " 'negative regulation of myeloid leukocyte differentiation',\n",
       " 'positive regulation of myeloid leukocyte differentiation',\n",
       " 'negative regulation of antimicrobial peptide production',\n",
       " 'negative regulation of antibacterial peptide production',\n",
       " 'negative regulation of antifungal peptide production',\n",
       " 'regulation of peptide secretion',\n",
       " 'negative regulation of peptide secretion',\n",
       " 'positive regulation of peptide secretion',\n",
       " 'regulation of antimicrobial peptide secretion',\n",
       " 'negative regulation of antimicrobial peptide secretion',\n",
       " 'positive regulation of antimicrobial peptide secretion',\n",
       " 'regulation of antibacterial peptide secretion',\n",
       " 'negative regulation of antibacterial peptide secretion',\n",
       " 'positive regulation of antibacterial peptide secretion',\n",
       " 'regulation of antifungal peptide secretion',\n",
       " 'negative regulation of antifungal peptide secretion',\n",
       " 'positive regulation of antifungal peptide secretion',\n",
       " 'positive regulation of antibacterial peptide production',\n",
       " 'positive regulation of antifungal peptide production',\n",
       " 'regulation of antimicrobial peptide biosynthetic process',\n",
       " 'negative regulation of antimicrobial peptide biosynthetic process',\n",
       " 'positive regulation of antimicrobial peptide biosynthetic process',\n",
       " 'regulation of antibacterial peptide biosynthetic process',\n",
       " 'negative regulation of antibacterial peptide biosynthetic process',\n",
       " 'regulation of antifungal peptide biosynthetic process',\n",
       " 'negative regulation of antifungal peptide biosynthetic process',\n",
       " 'regulation of biosynthetic process of antibacterial peptides active against Gram-negative bacteria',\n",
       " 'negative regulation of biosynthetic process of antibacterial peptides active against Gram-negative bacteria',\n",
       " 'regulation of biosynthetic process of antibacterial peptides active against Gram-positive bacteria',\n",
       " 'negative regulation of biosynthetic process of antibacterial peptides active against Gram-positive bacteria',\n",
       " 'negative regulation of adaptive immune response',\n",
       " 'positive regulation of adaptive immune response',\n",
       " 'regulation of adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains',\n",
       " 'negative regulation of adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains',\n",
       " 'positive regulation of adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains',\n",
       " 'regulation of T-helper 1 type immune response',\n",
       " 'negative regulation of T-helper 1 type immune response',\n",
       " 'positive regulation of T-helper 1 type immune response',\n",
       " 'negative regulation of type 2 immune response',\n",
       " 'positive regulation of type 2 immune response',\n",
       " 'negative regulation of response to tumor cell',\n",
       " 'positive regulation of response to tumor cell',\n",
       " 'negative regulation of immune response to tumor cell',\n",
       " 'positive regulation of immune response to tumor cell',\n",
       " 'regulation of T cell mediated immune response to tumor cell',\n",
       " 'negative regulation of T cell mediated immune response to tumor cell',\n",
       " 'positive regulation of T cell mediated immune response to tumor cell',\n",
       " 'regulation of tolerance induction to tumor cell',\n",
       " 'negative regulation of tolerance induction to tumor cell',\n",
       " 'positive regulation of tolerance induction to tumor cell',\n",
       " 'regulation of T cell tolerance induction to tumor cell',\n",
       " 'negative regulation of T cell tolerance induction to tumor cell',\n",
       " 'positive regulation of T cell tolerance induction to tumor cell',\n",
       " 'regulation of peripheral T cell tolerance induction',\n",
       " 'negative regulation of peripheral T cell tolerance induction',\n",
       " 'positive regulation of peripheral T cell tolerance induction',\n",
       " 'regulation of T cell mediated cytotoxicity directed against tumor cell target',\n",
       " 'negative regulation of T cell mediated cytotoxicity directed against tumor cell target',\n",
       " 'positive regulation of T cell mediated cytotoxicity directed against tumor cell target',\n",
       " 'regulation of natural killer cell mediated immune response to tumor cell',\n",
       " 'negative regulation of natural killer cell mediated immune response to tumor cell',\n",
       " 'positive regulation of natural killer cell mediated immune response to tumor cell',\n",
       " 'regulation of natural killer cell mediated cytotoxicity directed against tumor cell target',\n",
       " 'negative regulation of natural killer cell mediated cytotoxicity directed against tumor cell target',\n",
       " 'positive regulation of natural killer cell mediated cytotoxicity directed against tumor cell target',\n",
       " 'negative regulation of inflammatory response to antigenic stimulus',\n",
       " 'positive regulation of inflammatory response to antigenic stimulus',\n",
       " 'regulation of acute inflammatory response to antigenic stimulus',\n",
       " 'negative regulation of acute inflammatory response to antigenic stimulus',\n",
       " 'positive regulation of acute inflammatory response to antigenic stimulus',\n",
       " 'negative regulation of B cell deletion',\n",
       " 'positive regulation of B cell deletion',\n",
       " 'negative regulation of natural killer cell tolerance induction',\n",
       " 'positive regulation of natural killer cell tolerance induction',\n",
       " 'regulation of chronic inflammatory response to antigenic stimulus',\n",
       " 'negative regulation of chronic inflammatory response to antigenic stimulus',\n",
       " 'positive regulation of chronic inflammatory response to antigenic stimulus',\n",
       " 'regulation of acute inflammatory response to non-antigenic stimulus',\n",
       " 'negative regulation of acute inflammatory response to non-antigenic stimulus',\n",
       " 'positive regulation of acute inflammatory response to non-antigenic stimulus',\n",
       " 'regulation of chronic inflammatory response to non-antigenic stimulus',\n",
       " 'negative regulation of chronic inflammatory response to non-antigenic stimulus',\n",
       " 'positive regulation of chronic inflammatory response to non-antigenic stimulus',\n",
       " 'regulation of hypersensitivity',\n",
       " 'negative regulation of hypersensitivity',\n",
       " 'positive regulation of hypersensitivity',\n",
       " 'regulation of myeloid leukocyte mediated immunity',\n",
       " 'negative regulation of myeloid leukocyte mediated immunity',\n",
       " 'positive regulation of myeloid leukocyte mediated immunity',\n",
       " 'regulation of immunoglobulin mediated immune response',\n",
       " 'negative regulation of immunoglobulin mediated immune response',\n",
       " 'positive regulation of immunoglobulin mediated immune response',\n",
       " 'regulation of type II hypersensitivity',\n",
       " 'negative regulation of type II hypersensitivity',\n",
       " 'positive regulation of type II hypersensitivity',\n",
       " 'regulation of central B cell tolerance induction',\n",
       " 'negative regulation of central B cell tolerance induction',\n",
       " 'positive regulation of central B cell tolerance induction',\n",
       " 'regulation of central B cell deletion',\n",
       " 'negative regulation of central B cell deletion',\n",
       " 'positive regulation of central B cell deletion',\n",
       " 'regulation of B cell apoptotic process',\n",
       " 'negative regulation of B cell apoptotic process',\n",
       " 'positive regulation of B cell apoptotic process',\n",
       " 'regulation of peripheral B cell deletion',\n",
       " 'negative regulation of peripheral B cell deletion',\n",
       " 'positive regulation of peripheral B cell deletion',\n",
       " 'negative regulation of lymphocyte anergy',\n",
       " 'positive regulation of lymphocyte anergy',\n",
       " 'regulation of central B cell anergy',\n",
       " 'negative regulation of central B cell anergy',\n",
       " 'positive regulation of central B cell anergy',\n",
       " 'regulation of peripheral B cell anergy',\n",
       " 'negative regulation of peripheral B cell anergy',\n",
       " 'positive regulation of peripheral B cell anergy',\n",
       " 'negative regulation of humoral immune response',\n",
       " 'positive regulation of humoral immune response',\n",
       " 'regulation of humoral immune response mediated by circulating immunoglobulin',\n",
       " 'negative regulation of humoral immune response mediated by circulating immunoglobulin',\n",
       " 'positive regulation of humoral immune response mediated by circulating immunoglobulin',\n",
       " 'regulation of systemic arterial blood pressure by carotid body chemoreceptor signaling',\n",
       " 'regulation of systemic arterial blood pressure by aortic body chemoreceptor signaling',\n",
       " 'regulation of systemic arterial blood pressure by stress relaxation',\n",
       " 'regulation of systemic arterial blood pressure by epinephrine',\n",
       " 'regulation of systemic arterial blood pressure by norepinephrine',\n",
       " 'regulation of systemic arterial blood pressure by capillary fluid shift',\n",
       " 'regulation of vascular smooth muscle contraction',\n",
       " 'positive regulation of the force of heart contraction by epinephrine',\n",
       " 'negative regulation of the force of heart contraction by acetylcholine',\n",
       " 'positive regulation of the force of heart contraction by norepinephrine',\n",
       " 'regulation of systemic arterial blood pressure by acetylcholine',\n",
       " 'acetylcholine-mediated vasodilation involved in regulation of systemic arterial blood pressure',\n",
       " 'renal vasodilation of the peripheral vascular system involved in regulation of systemic arterial blood pressure',\n",
       " 'negative regulation of renal output by angiotensin',\n",
       " 'regulation of systemic arterial blood pressure by local renal renin-angiotensin',\n",
       " 'positive regulation of the force of heart contraction by neuronal epinephrine',\n",
       " 'positive regulation of the force of heart contraction by circulating epinephrine',\n",
       " 'positive regulation of the force of heart contraction by circulating epinephrine-norepinephrine',\n",
       " 'positive regulation of the force of heart contraction by neuronal epinephrine-norepinephrine',\n",
       " 'regulation of glomerular filtration',\n",
       " 'negative regulation of glomerular filtration by angiotensin',\n",
       " 'positive regulation of the force of heart contraction by circulating norepinephrine',\n",
       " 'positive regulation of the force of heart contraction by neuronal norepinephrine',\n",
       " 'positive regulation of heart rate by circulating epinephrine',\n",
       " 'positive regulation of heart rate by neuronal epinephrine',\n",
       " 'positive regulation of heart rate by neuronal norepinephrine',\n",
       " 'positive regulation of heart rate by circulating norepinephrine',\n",
       " 'regulation of vasoconstriction by epinephrine',\n",
       " 'regulation of vasoconstriction by norepinephrine',\n",
       " 'regulation of vasoconstriction by circulating norepinephrine',\n",
       " 'regulation of vasoconstriction by neuronal norepinephrine',\n",
       " 'regulation of vasoconstriction by neuronal epinephrine',\n",
       " 'regulation of vasoconstriction by circulating epinephrine',\n",
       " 'negative regulation of heart induction by canonical Wnt signaling pathway',\n",
       " 'regulation of animal organ formation',\n",
       " 'regulation of cell proliferation involved in heart valve morphogenesis',\n",
       " 'positive regulation of cell proliferation involved in heart valve morphogenesis',\n",
       " 'negative regulation of cell proliferation involved in heart valve morphogenesis',\n",
       " 'regulation of transcription from RNA polymerase II promoter involved in myocardial precursor cell differentiation',\n",
       " 'positive regulation of transcription from RNA polymerase II promoter involved in myocardial precursor cell differentiation',\n",
       " 'regulation of transcription from RNA polymerase II promoter involved in endocardial precursor cell differentiation',\n",
       " 'regulation of cardioblast proliferation',\n",
       " 'regulation of primary heart field cardioblast proliferation',\n",
       " 'regulation of secondary heart field cardioblast proliferation',\n",
       " 'canonical Wnt signaling pathway involved in positive regulation of secondary heart field cardioblast proliferation',\n",
       " 'fibroblast growth factor receptor signaling pathway involved in regulation of secondary heart field cardioblast proliferation',\n",
       " 'BMP signaling pathway involved in regulation of secondary heart field cardioblast proliferation',\n",
       " 'Notch signaling pathway involved in regulation of secondary heart field cardioblast proliferation',\n",
       " 'positive regulation of extracellular matrix constituent secretion',\n",
       " 'negative regulation of extracellular matrix constituent secretion',\n",
       " 'regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis',\n",
       " 'negative regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis',\n",
       " 'regulation of cilium movement',\n",
       " 'positive regulation of cilium movement',\n",
       " 'negative regulation of cilium movement',\n",
       " 'regulation of cilium beat frequency',\n",
       " 'regulation of dynamin family protein polymerization involved in membrane fission',\n",
       " 'regulation of glycogen biosynthetic process',\n",
       " 'regulation of glycogen catabolic process',\n",
       " 'regulation of gluconeogenesis',\n",
       " 'regulation of DNA replication',\n",
       " 'regulation of DNA repair',\n",
       " 'regulation of transcription, DNA-templated',\n",
       " 'regulation of transcription by RNA polymerase I',\n",
       " 'regulation of transcription by RNA polymerase II',\n",
       " 'regulation of transcription by RNA polymerase III',\n",
       " 'regulation of translation',\n",
       " 'regulation of translational initiation',\n",
       " 'regulation of translational initiation by iron',\n",
       " 'regulation of translational elongation',\n",
       " 'regulation of translational termination',\n",
       " 'negative regulation of protein kinase activity',\n",
       " 'regulation of pH',\n",
       " 'regulation of muscle contraction',\n",
       " 'regulation of smooth muscle contraction',\n",
       " 'regulation of striated muscle contraction',\n",
       " 'positive regulation of antibacterial peptide biosynthetic process',\n",
       " 'positive regulation of biosynthetic process of antibacterial peptides active against Gram-negative bacteria',\n",
       " 'positive regulation of biosynthetic process of antibacterial peptides active against Gram-positive bacteria',\n",
       " 'positive regulation of antifungal peptide biosynthetic process',\n",
       " 'positive regulation of transcription from RNA polymerase II promoter involved in unfolded protein response',\n",
       " 'negative regulation of microtubule depolymerization',\n",
       " 'negative regulation of axonemal microtubule depolymerization',\n",
       " 'regulation of sister chromatid cohesion',\n",
       " 'regulation of mitotic nuclear division',\n",
       " 'regulation of exit from mitosis',\n",
       " 'negative regulation of adenylate cyclase activity',\n",
       " 'positive regulation of cytosolic calcium ion concentration',\n",
       " 'positive regulation of transcription of Notch receptor target',\n",
       " 'negative regulation of oskar mRNA translation',\n",
       " 'positive regulation of transcription from RNA polymerase II promoter by pheromones',\n",
       " 'regulation of preblastoderm mitotic cell cycle',\n",
       " 'negative regulation of neuroblast proliferation',\n",
       " 'regulation of rhodopsin gene expression',\n",
       " 'regulation of ecdysteroid biosynthetic process',\n",
       " 'regulation of ecdysteroid secretion',\n",
       " 'regulation of juvenile hormone biosynthetic process',\n",
       " 'regulation of heart contraction',\n",
       " 'negative regulation of DNA replication',\n",
       " 'regulation of bicoid mRNA localization',\n",
       " 'regulation of Wnt signaling pathway, calcium modulating pathway',\n",
       " 'positive regulation of transcription by pheromones',\n",
       " 'regulation of transcription by pheromones',\n",
       " 'negative regulation of abscisic acid-activated signaling pathway',\n",
       " 'positive regulation of abscisic acid-activated signaling pathway',\n",
       " 'regulation of mitotic cell cycle, embryonic',\n",
       " 'negative regulation of gibberellic acid mediated signaling pathway',\n",
       " 'positive regulation of gibberellic acid mediated signaling pathway',\n",
       " 'positive regulation of flavonoid biosynthetic process',\n",
       " 'negative regulation of flavonoid biosynthetic process',\n",
       " 'negative regulation of cell fate specification',\n",
       " 'negative regulation of cardioblast cell fate specification',\n",
       " 'negative regulation of retinal cone cell fate specification',\n",
       " 'negative regulation of auditory receptor cell fate specification',\n",
       " 'regulation of seed germination',\n",
       " 'positive regulation of seed germination',\n",
       " 'regulation of atrichoblast fate specification',\n",
       " 'positive regulation of atrichoblast fate specification',\n",
       " 'negative regulation of atrichoblast fate specification',\n",
       " 'regulation of trichoblast fate specification',\n",
       " 'negative regulation of trichoblast fate specification',\n",
       " 'positive regulation of trichoblast fate specification',\n",
       " 'regulation of root meristem growth',\n",
       " 'regulation of vegetative meristem growth',\n",
       " 'regulation of ethylene-activated signaling pathway',\n",
       " 'negative regulation of ethylene-activated signaling pathway',\n",
       " 'regulation of photosynthesis, dark reaction',\n",
       " 'negative regulation of systemic acquired resistance',\n",
       " 'positive regulation of abscisic acid biosynthetic process',\n",
       " 'regulation of proton transport',\n",
       " 'regulation of cellular defense response',\n",
       " 'positive regulation of cellular defense response',\n",
       " 'negative regulation of seed germination',\n",
       " 'regulation of chlorophyll catabolic process',\n",
       " 'regulation of hydrogen peroxide metabolic process',\n",
       " 'regulation of vegetative phase change',\n",
       " 'regulation of isopentenyl diphosphate biosynthetic process, methylerythritol 4-phosphate pathway',\n",
       " 'negative regulation of isopentenyl diphosphate biosynthetic process, methylerythritol 4-phosphate pathway',\n",
       " 'regulation of anion channel activity',\n",
       " 'negative regulation of anion channel activity',\n",
       " 'regulation of anion channel activity by blue light',\n",
       " 'negative regulation of anion channel activity by blue light',\n",
       " 'regulation of ethylene biosynthetic process',\n",
       " 'positive regulation of ethylene biosynthetic process',\n",
       " 'negative regulation of ethylene biosynthetic process',\n",
       " 'positive regulation of gibberellin biosynthetic process',\n",
       " 'negative regulation of gibberellin biosynthetic process',\n",
       " 'regulation of chlorophyll biosynthetic process',\n",
       " 'regulation of G2/M transition of mitotic cell cycle',\n",
       " 'regulation of brassinosteroid biosynthetic process',\n",
       " 'negative regulation of brassinosteroid biosynthetic process',\n",
       " 'negative regulation of cell fate commitment',\n",
       " 'positive regulation of cell fate commitment',\n",
       " 'regulation of light-activated voltage-gated calcium channel activity',\n",
       " 'negative regulation of peptidase activity',\n",
       " 'regulation of autophagy',\n",
       " 'negative regulation of autophagy',\n",
       " 'positive regulation of autophagy',\n",
       " 'regulation of acetyl-CoA biosynthetic process from pyruvate',\n",
       " 'regulation of phosphatidylinositol biosynthetic process',\n",
       " 'negative regulation of phosphatidylinositol biosynthetic process',\n",
       " 'positive regulation of phosphatidylinositol biosynthetic process',\n",
       " 'negative regulation of induction of conjugation with cellular fusion',\n",
       " 'regulation of phospholipase activity',\n",
       " 'positive regulation of phospholipase activity',\n",
       " 'negative regulation of phospholipase activity',\n",
       " 'regulation of reciprocal meiotic recombination',\n",
       " 'negative regulation of transposition, RNA-mediated',\n",
       " 'positive regulation of transposition, RNA-mediated',\n",
       " 'regulation of activation of Janus kinase activity',\n",
       " 'positive regulation of activation of Janus kinase activity',\n",
       " 'negative regulation of platelet activation',\n",
       " 'regulation of thylakoid membrane disassembly',\n",
       " 'regulation of PSII associated light-harvesting complex II catabolic process',\n",
       " 'positive regulation of macromolecule biosynthetic process',\n",
       " 'negative regulation of macromolecule biosynthetic process',\n",
       " 'regulation of glycoprotein biosynthetic process',\n",
       " 'positive regulation of glycoprotein biosynthetic process',\n",
       " 'negative regulation of glycoprotein biosynthetic process',\n",
       " 'positive regulation of phosphorus metabolic process',\n",
       " 'negative regulation of phosphorus metabolic process',\n",
       " 'regulation of ketone biosynthetic process',\n",
       " 'regulation of ketone catabolic process',\n",
       " 'regulation of double-strand break repair via homologous recombination',\n",
       " 'positive regulation of nuclear cell cycle DNA replication',\n",
       " 'positive regulation of platelet activation',\n",
       " 'positive regulation of vascular endothelial growth factor production',\n",
       " 'regulation of adenylate cyclase activity involved in G protein-coupled receptor signaling pathway',\n",
       " 'regulation of starch biosynthetic process',\n",
       " 'regulation of lamellipodium assembly',\n",
       " 'positive regulation of lamellipodium assembly',\n",
       " 'negative regulation of lamellipodium assembly',\n",
       " 'positive regulation of endothelial cell migration',\n",
       " 'negative regulation of endothelial cell migration',\n",
       " 'regulation of auxin biosynthetic process',\n",
       " 'positive regulation of auxin biosynthetic process',\n",
       " 'regulation of 1-aminocyclopropane-1-carboxylate metabolic process',\n",
       " 'positive regulation of cytoplasmic mRNA processing body assembly',\n",
       " 'negative regulation of cytoplasmic mRNA processing body assembly',\n",
       " 'posttranscriptional regulation of gene expression',\n",
       " 'positive regulation of cardiac muscle hypertrophy',\n",
       " 'negative regulation of cardiac muscle hypertrophy',\n",
       " 'positive regulation of cardiac muscle adaptation',\n",
       " 'negative regulation of cardiac muscle adaptation',\n",
       " 'negative regulation of transcription by transcription factor catabolism',\n",
       " 'regulation of Schwann cell proliferation',\n",
       " 'positive regulation of Schwann cell proliferation',\n",
       " 'negative regulation of Schwann cell proliferation',\n",
       " 'positive regulation of gene expression',\n",
       " 'negative regulation of gene expression',\n",
       " 'regulation of transcription, start site selection',\n",
       " 'positive regulation of organelle organization',\n",
       " 'negative regulation of organelle organization',\n",
       " 'positive regulation of platelet-derived growth factor receptor signaling pathway',\n",
       " 'negative regulation of platelet-derived growth factor receptor signaling pathway',\n",
       " 'positive regulation of cell communication by electrical coupling',\n",
       " 'negative regulation of cell communication by electrical coupling',\n",
       " 'positive regulation of cell communication by chemical coupling',\n",
       " 'negative regulation of cell communication by chemical coupling',\n",
       " 'negative regulation of muscle cell apoptotic process',\n",
       " 'regulation of muscle cell apoptotic process',\n",
       " 'positive regulation of muscle cell apoptotic process',\n",
       " 'regulation of striated muscle cell apoptotic process',\n",
       " 'positive regulation of striated muscle cell apoptotic process',\n",
       " 'negative regulation of striated muscle cell apoptotic process',\n",
       " 'regulation of cardiac muscle cell apoptotic process',\n",
       " 'positive regulation of cardiac muscle cell apoptotic process',\n",
       " 'negative regulation of cardiac muscle cell apoptotic process',\n",
       " 'positive regulation of cellular carbohydrate metabolic process',\n",
       " 'negative regulation of cellular carbohydrate metabolic process',\n",
       " 'negative regulation of cellular carbohydrate metabolic process by negative regulation of transcription, DNA-templated',\n",
       " 'negative regulation of ribosomal protein gene transcription by RNA polymerase II',\n",
       " 'negative regulation of ribosomal protein gene transcription from RNA polymerase II promoter in response to chemical stimulus',\n",
       " 'regulation of alkaline phosphatase activity',\n",
       " 'negative regulation of alkaline phosphatase activity',\n",
       " 'positive regulation of alkaline phosphatase activity',\n",
       " 'regulation of mitotic spindle pole body separation',\n",
       " 'positive regulation of mitotic spindle pole body separation',\n",
       " 'negative regulation of mitotic spindle pole body separation',\n",
       " 'negative regulation of norepinephrine secretion',\n",
       " 'positive regulation of norepinephrine secretion',\n",
       " 'negative regulation of collagen catabolic process',\n",
       " 'regulation of extracellular matrix disassembly',\n",
       " 'negative regulation of extracellular matrix disassembly',\n",
       " 'positive regulation of cell development',\n",
       " 'negative regulation of cell development',\n",
       " 'positive regulation of transcription from RNA polymerase II promoter in response to iron',\n",
       " 'positive regulation of hydrogen peroxide metabolic process',\n",
       " 'negative regulation of hydrogen peroxide metabolic process',\n",
       " 'regulation of hydrogen peroxide biosynthetic process',\n",
       " 'positive regulation of hydrogen peroxide biosynthetic process',\n",
       " 'negative regulation of hydrogen peroxide biosynthetic process',\n",
       " 'regulation of protein glutathionylation',\n",
       " 'positive regulation of protein glutathionylation',\n",
       " 'negative regulation of protein glutathionylation',\n",
       " 'positive regulation of transcription via serum response element binding',\n",
       " 'regulation of protein kinase A signaling',\n",
       " 'positive regulation of protein kinase A signaling',\n",
       " 'positive regulation of macrophage derived foam cell differentiation',\n",
       " 'negative regulation of macrophage derived foam cell differentiation',\n",
       " 'regulation of long-chain fatty acid import across plasma membrane',\n",
       " 'positive regulation of long-chain fatty acid import across plasma membrane',\n",
       " 'negative regulation of long-chain fatty acid import across plasma membrane',\n",
       " 'regulation of nitric oxide mediated signal transduction',\n",
       " 'positive regulation of nitric oxide mediated signal transduction',\n",
       " 'negative regulation of nitric oxide mediated signal transduction',\n",
       " 'regulation of cGMP-mediated signaling',\n",
       " 'positive regulation of cGMP-mediated signaling',\n",
       " 'negative regulation of cGMP-mediated signaling',\n",
       " 'regulation of plasminogen activation',\n",
       " 'positive regulation of plasminogen activation',\n",
       " 'negative regulation of plasminogen activation',\n",
       " 'regulation of macrophage chemotaxis',\n",
       " 'positive regulation of macrophage chemotaxis',\n",
       " 'negative regulation of macrophage chemotaxis',\n",
       " 'regulation of fibroblast migration',\n",
       " 'positive regulation of fibroblast migration',\n",
       " 'negative regulation of fibroblast migration',\n",
       " 'positive regulation of sodium ion transport',\n",
       " 'negative regulation of sodium ion transport',\n",
       " 'negative regulation of transcription from RNA polymerase II promoter in response to UV-induced DNA damage',\n",
       " 'regulation of cell morphogenesis involved in differentiation',\n",
       " 'positive regulation of cell morphogenesis involved in differentiation',\n",
       " 'negative regulation of cell morphogenesis involved in differentiation',\n",
       " 'regulation of mRNA export from nucleus',\n",
       " 'regulation of dolichol biosynthetic process',\n",
       " 'regulation of ubiquinone biosynthetic process',\n",
       " 'regulation of multivesicular body size involved in endosome transport',\n",
       " 'regulation of multivesicular body size involved in ubiquitin-dependent protein catabolism',\n",
       " 'regulation of peptidyl-threonine phosphorylation',\n",
       " 'positive regulation of peptidyl-threonine phosphorylation',\n",
       " 'negative regulation of peptidyl-threonine phosphorylation',\n",
       " 'regulation of tumor necrosis factor-mediated signaling pathway',\n",
       " 'negative regulation of tumor necrosis factor-mediated signaling pathway',\n",
       " 'positive regulation of synaptic vesicle priming',\n",
       " 'negative regulation of synaptic vesicle priming',\n",
       " 'regulation of T cell chemotaxis',\n",
       " 'positive regulation of T cell chemotaxis',\n",
       " 'regulation of mitochondrion organization',\n",
       " 'positive regulation of mitochondrion organization',\n",
       " 'negative regulation of mitochondrion organization',\n",
       " 'regulation of centrosome duplication',\n",
       " 'positive regulation of centrosome duplication',\n",
       " 'negative regulation of centrosome duplication',\n",
       " 'regulation of glucose transmembrane transport',\n",
       " 'positive regulation of glucose transmembrane transport',\n",
       " 'negative regulation of glucose transmembrane transport',\n",
       " 'regulation of myotube differentiation',\n",
       " 'positive regulation of myotube differentiation',\n",
       " 'negative regulation of myotube differentiation',\n",
       " 'regulation of protein ADP-ribosylation',\n",
       " 'negative regulation of protein ADP-ribosylation',\n",
       " 'regulation of keratinocyte proliferation',\n",
       " 'positive regulation of keratinocyte proliferation',\n",
       " 'negative regulation of keratinocyte proliferation',\n",
       " 'positive regulation of reciprocal meiotic recombination',\n",
       " 'regulation of chromatin disassembly',\n",
       " 'regulation of proton-transporting ATPase activity, rotational mechanism',\n",
       " 'positive regulation of pathway-restricted SMAD protein phosphorylation',\n",
       " 'positive regulation of phospholipase C activity',\n",
       " 'positive regulation of protein histidine kinase activity',\n",
       " 'regulation of triglyceride biosynthetic process',\n",
       " 'positive regulation of triglyceride biosynthetic process',\n",
       " 'negative regulation of triglyceride biosynthetic process',\n",
       " 'positive regulation of receptor biosynthetic process',\n",
       " 'negative regulation of receptor biosynthetic process',\n",
       " 'positive regulation of cholesterol esterification',\n",
       " 'regulation of cholesterol efflux',\n",
       " 'positive regulation of cholesterol efflux',\n",
       " 'regulation of cardiac muscle contraction by regulation of the release of sequestered calcium ion',\n",
       " 'regulation of cardiac muscle contraction by calcium ion signaling',\n",
       " 'positive regulation of mitochondrial translation in response to stress',\n",
       " 'positive regulation of steroid biosynthetic process',\n",
       " 'negative regulation of steroid biosynthetic process',\n",
       " 'negative regulation of ergosterol biosynthetic process',\n",
       " 'regulation of triglyceride catabolic process',\n",
       " 'negative regulation of triglyceride catabolic process',\n",
       " 'positive regulation of triglyceride catabolic process',\n",
       " 'negative regulation of phosphatidylcholine catabolic process',\n",
       " 'regulation of UDP-glucose catabolic process',\n",
       " 'negative regulation of UDP-glucose catabolic process',\n",
       " 'positive regulation of glucose metabolic process',\n",
       " 'regulation of heparan sulfate proteoglycan biosynthetic process',\n",
       " 'positive regulation of heparan sulfate proteoglycan biosynthetic process',\n",
       " 'positive regulation of heparan sulfate proteoglycan biosynthesis by positive regulation of epimerase activity',\n",
       " 'regulation of sterigmatocystin biosynthetic process',\n",
       " 'positive regulation of sterigmatocystin biosynthetic process',\n",
       " 'negative regulation of very-low-density lipoprotein particle clearance',\n",
       " 'regulation of inositol phosphate biosynthetic process',\n",
       " 'negative regulation of inositol phosphate biosynthetic process',\n",
       " 'positive regulation of phosphatase activity',\n",
       " 'negative regulation of phosphatase activity',\n",
       " 'regulation of inositol-polyphosphate 5-phosphatase activity',\n",
       " 'positive regulation of inositol-polyphosphate 5-phosphatase activity',\n",
       " 'positive regulation of auxin mediated signaling pathway',\n",
       " 'negative regulation of auxin mediated signaling pathway',\n",
       " 'positive regulation of macrophage tolerance induction',\n",
       " 'regulation of macrophage cytokine production',\n",
       " 'negative regulation of macrophage cytokine production',\n",
       " 'regulation of cytoplasmic microtubule depolymerization',\n",
       " 'positive regulation of necrotic cell death',\n",
       " 'negative regulation of transcription by competitive promoter binding',\n",
       " 'regulation of meiotic joint molecule formation',\n",
       " 'negative regulation of meiotic joint molecule formation',\n",
       " 'negative regulation of cell cycle process',\n",
       " 'negative regulation of intestinal phytosterol absorption',\n",
       " 'positive regulation of endopeptidase activity',\n",
       " 'negative regulation of endopeptidase activity',\n",
       " 'positive regulation of peptidase activity',\n",
       " 'positive regulation of protein processing',\n",
       " 'negative regulation of protein processing',\n",
       " 'negative regulation of calcidiol 1-monooxygenase activity',\n",
       " 'negative regulation of vitamin D biosynthetic process',\n",
       " 'regulation of amino acid import across plasma membrane',\n",
       " 'regulation of metal ion transport',\n",
       " 'regulation of glucan biosynthetic process',\n",
       " 'regulation of chromatin silencing by small RNA',\n",
       " 'regulation of mitotic sister chromatid separation',\n",
       " 'regulation of phosphate transport',\n",
       " 'regulation of microtubule nucleation',\n",
       " 'regulation of pheromone-dependent signal transduction involved in conjugation with cellular fusion',\n",
       " 'positive regulation of G2/M transition of mitotic cell cycle',\n",
       " 'negative regulation of G2/M transition of mitotic cell cycle',\n",
       " 'positive regulation of division septum assembly',\n",
       " 'negative regulation of division septum assembly',\n",
       " 'regulation of neuron projection development',\n",
       " 'positive regulation of neuron projection development',\n",
       " 'negative regulation of neuron projection development',\n",
       " 'positive regulation of vitamin D 24-hydroxylase activity',\n",
       " 'positive regulation of high-density lipoprotein particle clearance',\n",
       " 'negative regulation of high-density lipoprotein particle clearance',\n",
       " 'negative regulation of low-density lipoprotein particle clearance',\n",
       " 'negative regulation of SMAD protein complex assembly',\n",
       " 'regulation of translational initiation by eIF2 alpha phosphorylation',\n",
       " 'regulation of microglia differentiation',\n",
       " 'negative regulation of microglia differentiation',\n",
       " 'positive regulation of microglia differentiation',\n",
       " 'regulation of gliogenesis',\n",
       " 'negative regulation of gliogenesis',\n",
       " 'positive regulation of gliogenesis',\n",
       " 'regulation of Schwann cell differentiation',\n",
       " 'negative regulation of Schwann cell differentiation',\n",
       " 'positive regulation of Schwann cell differentiation',\n",
       " 'regulation of neuron maturation',\n",
       " 'positive regulation of neuron maturation',\n",
       " 'negative regulation of neuron maturation',\n",
       " 'positive regulation of glutamate secretion',\n",
       " 'negative regulation of glutamate secretion',\n",
       " 'regulation of gamma-aminobutyric acid secretion',\n",
       " 'negative regulation of gamma-aminobutyric acid secretion',\n",
       " 'positive regulation of gamma-aminobutyric acid secretion',\n",
       " 'positive regulation of acetylcholine secretion, neurotransmission',\n",
       " 'negative regulation of acetylcholine secretion, neurotransmission',\n",
       " 'regulation of epinephrine secretion',\n",
       " 'regulation of norepinephrine secretion',\n",
       " 'negative regulation of serotonin secretion',\n",
       " 'positive regulation of serotonin secretion',\n",
       " 'regulation of phosphatidylinositol 3-kinase signaling',\n",
       " 'negative regulation of phosphatidylinositol 3-kinase signaling',\n",
       " 'positive regulation of phosphatidylinositol 3-kinase signaling',\n",
       " 'regulation of somitomeric trunk muscle development',\n",
       " 'positive regulation of somitomeric trunk muscle development',\n",
       " 'negative regulation of somitomeric trunk muscle development',\n",
       " 'regulation of branchiomeric skeletal muscle development',\n",
       " 'positive regulation of branchiomeric skeletal muscle development',\n",
       " 'negative regulation of branchiomeric skeletal muscle development',\n",
       " 'positive regulation of satellite cell activation involved in skeletal muscle regeneration',\n",
       " 'regulation of skeletal muscle contraction by calcium ion signaling',\n",
       " 'regulation of skeletal muscle contraction by modulation of calcium ion sensitivity of myofibril',\n",
       " 'regulation of twitch skeletal muscle contraction',\n",
       " 'regulation of extraocular skeletal muscle development',\n",
       " 'negative regulation of extraocular skeletal muscle development',\n",
       " 'positive regulation of extraocular skeletal muscle development',\n",
       " 'regulation of the force of skeletal muscle contraction',\n",
       " 'negative regulation of muscle atrophy',\n",
       " 'positive regulation of muscle atrophy',\n",
       " 'positive regulation of muscle hyperplasia',\n",
       " 'negative regulation of muscle hyperplasia',\n",
       " 'regulation of muscle hypertrophy',\n",
       " 'regulation of tonic skeletal muscle contraction',\n",
       " 'positive regulation of tonic skeletal muscle contraction',\n",
       " 'negative regulation of tonic skeletal muscle contraction',\n",
       " 'regulation of skeletal muscle contraction by regulation of release of sequestered calcium ion',\n",
       " 'positive regulation of skeletal muscle contraction by regulation of release of sequestered calcium ion',\n",
       " 'negative regulation of skeletal muscle contraction by regulation of release of sequestered calcium ion',\n",
       " 'regulation of skeletal muscle contraction',\n",
       " 'regulation of skeletal muscle satellite cell proliferation',\n",
       " 'growth factor dependent regulation of skeletal muscle satellite cell proliferation',\n",
       " 'regulation of skeletal muscle contraction by neural stimulation via neuromuscular junction',\n",
       " 'positive regulation of skeletal muscle cell proliferation',\n",
       " 'negative regulation of skeletal muscle cell proliferation',\n",
       " 'regulation of skeletal muscle contraction by chemo-mechanical energy conversion',\n",
       " 'cross bridge cycling involved in regulation of the velocity of shortening in skeletal muscle contraction',\n",
       " 'cross bridge formation involved in regulation of the velocity of shortening in skeletal muscle contraction',\n",
       " 'regulation of muscle filament sliding involved in regulation of the velocity of shortening in skeletal muscle contraction',\n",
       " 'regulation of smooth muscle cell migration',\n",
       " 'positive regulation of smooth muscle cell migration',\n",
       " 'negative regulation of smooth muscle cell migration',\n",
       " 'regulation of light-activated channel activity',\n",
       " 'positive regulation of macroautophagy',\n",
       " 'regulation of macroautophagy',\n",
       " 'negative regulation of macroautophagy',\n",
       " 'negative regulation of transcription by RNA polymerase I',\n",
       " 'negative regulation of transcription by RNA polymerase III',\n",
       " 'negative regulation of angiogenesis',\n",
       " 'regulation of transforming growth factor beta receptor signaling pathway',\n",
       " 'negative regulation of RNA polymerase II transcriptional preinitiation complex assembly',\n",
       " 'negative regulation of translation',\n",
       " 'calcium-ion regulated exocytosis',\n",
       " 'regulation of calcium ion-dependent exocytosis',\n",
       " 'modification by virus of host cell cycle regulation',\n",
       " 'regulation of fatty acid metabolic process',\n",
       " 'regulation of steroid metabolic process',\n",
       " 'regulation of phosphate metabolic process',\n",
       " 'regulation of vasoconstriction',\n",
       " 'negative regulation of cell-glial cell adhesion involved in cerebral cortex lamination',\n",
       " 'negative regulation of cell motility involved in cerebral cortex radial glia guided migration',\n",
       " 'regulation of transcription from RNA polymerase II promoter involved in forebrain neuron fate commitment',\n",
       " 'regulation of transcription from RNA polymerase II promoter involved in spinal cord motor neuron fate specification',\n",
       " 'regulation of transcription from RNA polymerase II promoter involved in ventral spinal cord interneuron specification',\n",
       " 'negative regulation of smoothened signaling pathway involved in ventral spinal cord patterning',\n",
       " 'regulation of transcription from RNA polymerase II promoter involved in somatic motor neuron fate commitment',\n",
       " 'regulation of transcription from RNA polymerase II promoter involved in spinal cord association neuron specification',\n",
       " 'regulation of cell proliferation in dorsal spinal cord',\n",
       " 'Wnt signaling pathway involved in regulation of cell proliferation in dorsal spinal cord',\n",
       " 'regulation of cerebellar granule cell precursor proliferation',\n",
       " 'positive regulation of cerebellar granule cell precursor proliferation',\n",
       " 'negative regulation of cerebellar granule cell precursor proliferation',\n",
       " 'positive regulation of cerebellar granule cell migration by calcium',\n",
       " 'negative regulation of anterior neural cell fate commitment of the neural plate',\n",
       " 'negative regulation of anterior neural cell fate commitment of the neural plate by Wnt signaling pathway',\n",
       " 'negative regulation of anterior neural cell fate commitment of the neural plate by fibroblast growth factor receptor signaling pathway',\n",
       " 'regulation of mitotic metaphase/anaphase transition',\n",
       " 'regulation of proteolysis',\n",
       " 'regulation of DNA-dependent DNA replication initiation',\n",
       " 'positive regulation of Wnt signaling pathway',\n",
       " 'negative regulation of Wnt signaling pathway',\n",
       " 'regulation of intestinal cholesterol absorption',\n",
       " 'positive regulation of cell migration',\n",
       " 'negative regulation of cell migration',\n",
       " 'regulation of complement activation, classical pathway',\n",
       " 'positive regulation of bone mineralization',\n",
       " 'negative regulation of bone mineralization',\n",
       " 'positive regulation of transforming growth factor beta receptor signaling pathway',\n",
       " 'negative regulation of transforming growth factor beta receptor signaling pathway',\n",
       " 'positive regulation of BMP signaling pathway',\n",
       " 'negative regulation of BMP signaling pathway',\n",
       " 'negative regulation of axon extension',\n",
       " 'regulation of cellular pH',\n",
       " 'positive regulation of border follicle cell delamination',\n",
       " 'regulation of cyclic nucleotide catabolic process',\n",
       " 'regulation of nucleotide biosynthetic process',\n",
       " 'negative regulation of nucleotide biosynthetic process',\n",
       " 'positive regulation of nucleotide biosynthetic process',\n",
       " 'regulation of nucleotide catabolic process',\n",
       " 'negative regulation of nucleotide catabolic process',\n",
       " 'positive regulation of nucleotide catabolic process',\n",
       " 'regulation of actin filament polymerization',\n",
       " 'regulation of actin filament depolymerization',\n",
       " 'negative regulation of actin filament depolymerization',\n",
       " 'positive regulation of actin filament depolymerization',\n",
       " 'negative regulation of actin filament polymerization',\n",
       " 'positive regulation of actin filament polymerization',\n",
       " 'regulation of intermediate filament polymerization',\n",
       " 'negative regulation of intermediate filament polymerization',\n",
       " 'positive regulation of intermediate filament polymerization',\n",
       " 'regulation of intermediate filament depolymerization',\n",
       " 'negative regulation of intermediate filament depolymerization',\n",
       " 'positive regulation of intermediate filament depolymerization',\n",
       " 'regulation of granulocyte differentiation',\n",
       " 'negative regulation of granulocyte differentiation',\n",
       " 'positive regulation of granulocyte differentiation',\n",
       " 'negative regulation of epithelial cell differentiation',\n",
       " 'positive regulation of epithelial cell differentiation',\n",
       " 'regulation of polarized epithelial cell differentiation',\n",
       " 'negative regulation of polarized epithelial cell differentiation',\n",
       " 'positive regulation of polarized epithelial cell differentiation',\n",
       " 'negative regulation of myeloid dendritic cell activation',\n",
       " 'positive regulation of myeloid dendritic cell activation',\n",
       " 'regulation of B cell proliferation',\n",
       " 'negative regulation of B cell proliferation',\n",
       " 'positive regulation of B cell proliferation',\n",
       " 'negative regulation of vascular endothelial growth factor receptor signaling pathway',\n",
       " 'positive regulation of vascular endothelial growth factor receptor signaling pathway',\n",
       " 'regulation of centriole-centriole cohesion',\n",
       " 'negative regulation of septation initiation signaling',\n",
       " 'positive regulation of septation initiation signaling',\n",
       " 'regulation of histone modification',\n",
       " 'negative regulation of histone modification',\n",
       " 'positive regulation of histone modification',\n",
       " 'regulation of histone methylation',\n",
       " 'negative regulation of histone methylation',\n",
       " 'positive regulation of histone methylation',\n",
       " 'regulation of histone deacetylation',\n",
       " 'negative regulation of histone deacetylation',\n",
       " 'positive regulation of histone deacetylation',\n",
       " 'regulation of histone deacetylation at centromere',\n",
       " 'negative regulation of histone deacetylation at centromere',\n",
       " 'positive regulation of histone deacetylation at centromere',\n",
       " 'regulation of microtubule polymerization or depolymerization',\n",
       " 'negative regulation of microtubule polymerization or depolymerization',\n",
       " 'positive regulation of microtubule polymerization or depolymerization',\n",
       " 'regulation of microtubule polymerization',\n",
       " 'regulation of microtubule depolymerization',\n",
       " 'negative regulation of microtubule polymerization',\n",
       " 'positive regulation of microtubule polymerization',\n",
       " 'positive regulation of microtubule depolymerization',\n",
       " 'regulation of pseudopodium assembly',\n",
       " 'negative regulation of pseudopodium assembly',\n",
       " 'positive regulation of pseudopodium assembly',\n",
       " 'regulation of lateral pseudopodium assembly',\n",
       " 'negative regulation of lateral pseudopodium assembly',\n",
       " 'positive regulation of lateral pseudopodium assembly',\n",
       " 'negative regulation of guanylate cyclase activity',\n",
       " 'positive regulation of guanylate cyclase activity',\n",
       " 'negative regulation of cellular biosynthetic process',\n",
       " 'positive regulation of cellular biosynthetic process',\n",
       " 'negative regulation of cellular catabolic process',\n",
       " 'positive regulation of cellular catabolic process',\n",
       " 'regulation of sulfur amino acid metabolic process',\n",
       " 'negative regulation of sulfur amino acid metabolic process',\n",
       " 'positive regulation of sulfur amino acid metabolic process',\n",
       " 'negative regulation of cell projection organization',\n",
       " 'positive regulation of cell projection organization',\n",
       " 'regulation of mating projection assembly',\n",
       " 'regulation of initiation of mating projection growth',\n",
       " 'regulation of termination of mating projection growth',\n",
       " 'regulation of prostaglandin biosynthetic process',\n",
       " 'negative regulation of prostaglandin biosynthetic process',\n",
       " 'positive regulation of prostaglandin biosynthetic process',\n",
       " 'regulation of protein ubiquitination',\n",
       " 'negative regulation of protein ubiquitination',\n",
       " 'positive regulation of protein ubiquitination',\n",
       " 'regulation of protein modification process',\n",
       " 'negative regulation of protein modification process',\n",
       " 'positive regulation of protein modification process',\n",
       " 'regulation of mRNA cleavage',\n",
       " 'negative regulation of mRNA cleavage',\n",
       " 'positive regulation of mRNA cleavage',\n",
       " \"regulation of mRNA 3'-end processing\",\n",
       " \"negative regulation of mRNA 3'-end processing\",\n",
       " \"positive regulation of mRNA 3'-end processing\",\n",
       " 'regulation of fast-twitch skeletal muscle fiber contraction',\n",
       " 'negative regulation of fast-twitch skeletal muscle fiber contraction',\n",
       " 'positive regulation of fast-twitch skeletal muscle fiber contraction',\n",
       " 'regulation of slow-twitch skeletal muscle fiber contraction',\n",
       " 'negative regulation of slow-twitch skeletal muscle fiber contraction',\n",
       " 'positive regulation of slow-twitch skeletal muscle fiber contraction',\n",
       " 'negative regulation of heterochromatin assembly',\n",
       " 'regulation of extent of heterochromatin assembly',\n",
       " 'positive regulation of exit from mitosis',\n",
       " 'regulation of anthocyanin biosynthetic process',\n",
       " 'negative regulation of anthocyanin biosynthetic process',\n",
       " 'positive regulation of anthocyanin biosynthetic process',\n",
       " 'negative regulation of brain-derived neurotrophic factor receptor signaling pathway',\n",
       " 'positive regulation of brain-derived neurotrophic factor receptor signaling pathway',\n",
       " 'regulation of DNA-templated transcription, termination',\n",
       " 'negative regulation of inositol 1,4,5-trisphosphate-sensitive calcium-release channel activity',\n",
       " 'positive regulation of inositol 1,4,5-trisphosphate-sensitive calcium-release channel activity',\n",
       " 'regulation of synaptic vesicle fusion to presynaptic active zone membrane',\n",
       " 'negative regulation of synaptic vesicle fusion to presynaptic active zone membrane',\n",
       " 'positive regulation of synaptic vesicle fusion to presynaptic active zone membrane',\n",
       " 'regulation of cyclin-dependent protein serine/threonine kinase activity involved in G2/M transition of mitotic cell cycle',\n",
       " 'negative regulation of cyclin-dependent protein serine/threonine kinase activity involved in G2/M transition of mitotic cell cycle',\n",
       " 'negative regulation of chromatin silencing',\n",
       " 'positive regulation of chromatin silencing',\n",
       " 'regulation of chromatin silencing at telomere',\n",
       " 'negative regulation of chromatin silencing at telomere',\n",
       " 'positive regulation of chromatin silencing at telomere',\n",
       " 'negative regulation of glucocorticoid metabolic process',\n",
       " 'positive regulation of glucocorticoid metabolic process',\n",
       " 'regulation of glucocorticoid biosynthetic process',\n",
       " 'negative regulation of glucocorticoid biosynthetic process',\n",
       " 'positive regulation of glucocorticoid biosynthetic process',\n",
       " 'regulation of glucocorticoid catabolic process',\n",
       " 'negative regulation of glucocorticoid catabolic process',\n",
       " ...]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[go.name for go in go_dag.values() if 'regula' in go.name and go.level > 5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conversion dictionaries\n",
    "GO2genes = pd.Series({go_id: set(genes.Systematic_ID) for go_id, genes in annotation_df.groupby('GO_ID')}, \n",
    "                     name='nb_genes')\n",
    "gene2GO = defaultdict(set)\n",
    "gene2GO  = {gene : set(go_ids.GO_ID)        for gene, go_ids in annotation_df.groupby('Systematic_ID')}\n",
    "global_GO_counter = GO2genes.apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_number_of_pre_runs(PVALUE_DIRECTORY, n_clusters = 99):\n",
    "    splitted_file_names = [name.split('_') for name in os.listdir(PVALUE_DIRECTORY)]\n",
    "    pre_runs = [int(run) for run, ncluster, db_txt in splitted_file_names if ncluster == str(n_clusters)]\n",
    "    if pre_runs:\n",
    "        return max(pre_runs)+1\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enrichments(alpha, p_values, cluster_list, correction):\n",
    "    relevant_p_values = [p_values[str(cluster_idx)][cluster2GO(cluster)] \n",
    "                             for cluster_idx,cluster in enumerate(cluster_list)] \n",
    "    \n",
    "    sorted_p_values = sorted(p for p_cluster in relevant_p_values\n",
    "                               for p in p_cluster)\n",
    "    m = len(sorted_p_values)\n",
    "    if   correction == 'BY':\n",
    "        c = np.log(m) + np.euler_gamma + 1/(2*m)\n",
    "    elif correction == 'BH':\n",
    "        c = 1\n",
    "    else:\n",
    "        print(\"Correction not known!\")\n",
    "        raise Exception\n",
    "    for k,P_k in enumerate(sorted_p_values,1):\n",
    "        if P_k > k/(m*c) * alpha:\n",
    "            break\n",
    "    threshold = sorted_p_values[k-2]\n",
    "    return p_values < threshold\n",
    "\n",
    "\n",
    "def cluster2GO(cluster):\n",
    "    return set.union(*(gene2GO.get(gene, set()) for gene in cluster))\n",
    "\n",
    "def is_annotated_in(gene, GO_set):\n",
    "    return not gene2GO.get(gene,set()).isdisjoint(GO_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages = {}\n",
    "GO_coverages      = {}\n",
    "gene_coverages    = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "MIN_CLUSTERS = 2\n",
    "MAX_CLUSTERS = 99\n",
    "MAX_RUNS     = 30"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GDV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = 'GDV'\n",
    "\n",
    "all_distances = sorted('_'.join(filename.split('_')[:-1]) \n",
    "                           for filename in os.listdir(f\"{MATRIX_DIRECTORY}/{feature}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages['GDV'] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     'GDV'] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   'GDV'] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDV_similarity 0\n",
      "GDV_similarity 1\n",
      "GDV_similarity 2\n",
      "GDV_similarity 3\n",
      "GDV_similarity 4\n",
      "GDV_similarity 5\n",
      "GDV_similarity 6\n",
      "GDV_similarity 7\n",
      "GDV_similarity 8\n",
      "GDV_similarity 9\n",
      "98: 11.25sec\n",
      "braycurtis 0\n",
      "braycurtis 1\n",
      "braycurtis 2\n",
      "braycurtis 3\n",
      "braycurtis 4\n",
      "braycurtis 5\n",
      "braycurtis 6\n",
      "braycurtis 7\n",
      "braycurtis 8\n",
      "braycurtis 9\n",
      "98: 11.47sec\n",
      "canberra 0\n",
      "canberra 1ec\n",
      "canberra 2ec\n",
      "canberra 3ec\n",
      "canberra 4ec\n",
      "canberra 5ec\n",
      "canberra 6ec\n",
      "canberra 7ec\n",
      "canberra 8ec\n",
      "canberra 9ec\n",
      "98: 11.57sec\n",
      "chebyshev 0\n",
      "chebyshev 1c\n",
      "chebyshev 2c\n",
      "chebyshev 3c\n",
      "chebyshev 4c\n",
      "chebyshev 5c\n",
      "chebyshev 6c\n",
      "chebyshev 7c\n",
      "chebyshev 8c\n",
      "chebyshev 9c\n",
      "98: 16.06sec\n",
      "cityblock 0\n",
      "cityblock 1c\n",
      "cityblock 2c\n",
      "cityblock 3c\n",
      "cityblock 4c\n",
      "cityblock 5c\n",
      "cityblock 6c\n",
      "cityblock 7c\n",
      "cityblock 8c\n",
      "cityblock 9c\n",
      "98: 17.34sec\n",
      "correlation 0\n",
      "correlation 1\n",
      "correlation 2\n",
      "correlation 3\n",
      "correlation 4\n",
      "correlation 5\n",
      "correlation 6\n",
      "correlation 7\n",
      "correlation 8\n",
      "correlation 9\n",
      "98: 18.78sec\n",
      "cosine 0\n",
      "cosine 17sec\n",
      "cosine 22sec\n",
      "cosine 36sec\n",
      "cosine 49sec\n",
      "cosine 58sec\n",
      "cosine 60sec\n",
      "cosine 71sec\n",
      "cosine 84sec\n",
      "cosine 95sec\n",
      "98: 15.65sec\n",
      "euclidean 0\n",
      "euclidean 1c\n",
      "euclidean 2c\n",
      "euclidean 3c\n",
      "euclidean 4c\n",
      "euclidean 5c\n",
      "euclidean 6c\n",
      "euclidean 7c\n",
      "euclidean 8c\n",
      "euclidean 9c\n",
      "98: 13.99sec\n",
      "mahalanobis 0\n",
      "mahalanobis 1\n",
      "mahalanobis 2\n",
      "mahalanobis 3\n",
      "mahalanobis 4\n",
      "mahalanobis 5\n",
      "mahalanobis 6\n",
      "mahalanobis 7\n",
      "mahalanobis 8\n",
      "mahalanobis 9\n",
      "98: 12.16sec\n",
      "normalized1_l1 0\n",
      "normalized1_l1 1\n",
      "normalized1_l1 2\n",
      "normalized1_l1 3\n",
      "normalized1_l1 4\n",
      "normalized1_l1 5\n",
      "normalized1_l1 6\n",
      "normalized1_l1 7\n",
      "normalized1_l1 8\n",
      "normalized1_l1 9\n",
      "98: 12.88sec\n",
      "normalized1_l2 0\n",
      "normalized1_l2 1\n",
      "normalized1_l2 2\n",
      "normalized1_l2 3\n",
      "normalized1_l2 4\n",
      "normalized1_l2 5\n",
      "normalized1_l2 6\n",
      "normalized1_l2 7\n",
      "normalized1_l2 8\n",
      "normalized1_l2 9\n",
      "98: 17.67sec\n",
      "normalized1_linf 0\n",
      "normalized1_linf 1\n",
      "normalized1_linf 2\n",
      "normalized1_linf 3\n",
      "normalized1_linf 4\n",
      "normalized1_linf 5\n",
      "normalized1_linf 6\n",
      "normalized1_linf 7\n",
      "normalized1_linf 8\n",
      "normalized1_linf 9\n",
      "98: 12.35sec\n",
      "normalized2_l1 0\n",
      "normalized2_l1 1\n",
      "normalized2_l1 2\n",
      "normalized2_l1 3\n",
      "normalized2_l1 4\n",
      "normalized2_l1 5\n",
      "normalized2_l1 6\n",
      "normalized2_l1 7\n",
      "normalized2_l1 8\n",
      "normalized2_l1 9\n",
      "98: 14.76sec\n",
      "normalized2_l2 0\n",
      "normalized2_l2 1\n",
      "normalized2_l2 2\n",
      "normalized2_l2 3\n",
      "normalized2_l2 4\n",
      "normalized2_l2 5\n",
      "normalized2_l2 6\n",
      "normalized2_l2 7\n",
      "normalized2_l2 8\n",
      "normalized2_l2 9\n",
      "98: 14.07sec\n",
      "normalized2_linf 0\n",
      "normalized2_linf 1\n",
      "normalized2_linf 2\n",
      "normalized2_linf 3\n",
      "normalized2_linf 4\n",
      "normalized2_linf 5\n",
      "normalized2_linf 6\n",
      "normalized2_linf 7\n",
      "normalized2_linf 8\n",
      "normalized2_linf 9\n",
      "98: 13.41sec\n",
      "seuclidean 0\n",
      "seuclidean 1\n",
      "seuclidean 2\n",
      "seuclidean 3\n",
      "seuclidean 4\n",
      "seuclidean 5\n",
      "seuclidean 6\n",
      "seuclidean 7\n",
      "seuclidean 8\n",
      "seuclidean 9\n",
      "98: 13.04sec\n",
      "sqeuclidean 0\n",
      "sqeuclidean 1\n",
      "sqeuclidean 2\n",
      "sqeuclidean 3\n",
      "sqeuclidean 4\n",
      "sqeuclidean 5\n",
      "sqeuclidean 6\n",
      "sqeuclidean 7\n",
      "sqeuclidean 8\n",
      "sqeuclidean 9\n",
      "98: 12.48sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in all_distances:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/GDV/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/GDV/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages['GDV'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS)) \n",
    "        GO_coverages[     'GDV'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS))\n",
    "        gene_coverages[   'GDV'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages['GDV'][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     'GDV'][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   'GDV'][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/GDV/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages['GDV'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     'GDV'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   'GDV'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GCV-A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages['GCV-A'] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     'GCV-A'] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   'GCV-A'] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canberra 0\n",
      "canberra 1ec\n",
      "canberra 2ec\n",
      "canberra 3ec\n",
      "canberra 4ec\n",
      "canberra 5ec\n",
      "canberra 6ec\n",
      "canberra 7ec\n",
      "canberra 8ec\n",
      "canberra 9ec\n",
      "canberra 10c\n",
      "canberra 11c\n",
      "canberra 12c\n",
      "canberra 13c\n",
      "canberra 14c\n",
      "canberra 15c\n",
      "canberra 16c\n",
      "canberra 17c\n",
      "canberra 18c\n",
      "canberra 19c\n",
      "canberra 20c\n",
      "canberra 21c\n",
      "canberra 22c\n",
      "canberra 23c\n",
      "canberra 24c\n",
      "canberra 25c\n",
      "canberra 26c\n",
      "canberra 27c\n",
      "canberra 28c\n",
      "canberra 29c\n",
      "98: 11.34sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in ['canberra']:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/GCV-A/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/GCV-A/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages['GCV-A'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     'GCV-A'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   'GCV-A'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages['GCV-A'][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     'GCV-A'][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   'GCV-A'][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/GCV-A/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages['GCV-A'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     'GCV-A'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   'GCV-A'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GCV-G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages['GCV-G'] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     'GCV-G'] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   'GCV-G'] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canberra 0\n",
      "canberra 1ec\n",
      "canberra 2ec\n",
      "canberra 3ec\n",
      "canberra 4ec\n",
      "canberra 5ec\n",
      "canberra 6ec\n",
      "canberra 7ec\n",
      "canberra 8ec\n",
      "canberra 9ec\n",
      "canberra 10c\n",
      "canberra 11c\n",
      "canberra 12c\n",
      "canberra 13c\n",
      "canberra 14c\n",
      "canberra 15c\n",
      "canberra 16c\n",
      "canberra 17c\n",
      "canberra 18c\n",
      "canberra 19c\n",
      "canberra 20c\n",
      "canberra 21c\n",
      "canberra 22c\n",
      "canberra 23c\n",
      "canberra 24c\n",
      "canberra 25c\n",
      "canberra 26c\n",
      "canberra 27c\n",
      "canberra 28c\n",
      "canberra 29c\n",
      "98: 11.29sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in ['canberra']:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/GCV-G/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/GCV-G/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages['GCV-G'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     'GCV-G'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   'GCV-G'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages['GCV-G'][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     'GCV-G'][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   'GCV-G'][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/GCV-G/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages['GCV-G'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     'GCV-G'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   'GCV-G'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GCV-DG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = 'GCV-DG'\n",
    "\n",
    "all_distances = sorted('_'.join(filename.split('_')[:-1]) \n",
    "                           for filename in os.listdir(f\"{MATRIX_DIRECTORY}/{feature}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages[feature] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     feature] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   feature] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canberra 0\n",
      "canberra 1ec\n",
      "canberra 2ec\n",
      "canberra 3ec\n",
      "canberra 4ec\n",
      "canberra 5ec\n",
      "canberra 6ec\n",
      "canberra 7ec\n",
      "canberra 8ec\n",
      "canberra 9ec\n",
      "98: 11.56sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in {'canberra'}:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/{feature}/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/{feature}/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages[feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages[feature][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     feature][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   feature][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/{feature}/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages[feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GCV-AD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages['GCV-AD'] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     'GCV-AD'] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   'GCV-AD'] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canberra 0\n",
      "canberra 1ec\n",
      "canberra 2ec\n",
      "canberra 3ec\n",
      "canberra 4ec\n",
      "canberra 5ec\n",
      "canberra 6ec\n",
      "canberra 7ec\n",
      "canberra 8ec\n",
      "canberra 9ec\n",
      "98: 15.54sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in ['canberra']:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/GCV-AD/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/GCV-AD/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages['GCV-AD'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     'GCV-AD'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   'GCV-AD'][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages['GCV-AD'][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     'GCV-AD'][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   'GCV-AD'][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/GCV-AD/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages['GCV-AD'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     'GCV-AD'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   'GCV-AD'][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GCV-all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = 'GCV-all'\n",
    "\n",
    "all_distances = sorted('_'.join(filename.split('_')[:-1]) \n",
    "                           for filename in os.listdir(f\"{MATRIX_DIRECTORY}/{feature}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages[feature] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     feature] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   feature] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canberra 0\n",
      "canberra 1ec\n",
      "canberra 2ec\n",
      "canberra 3ec\n",
      "canberra 4ec\n",
      "98: 11.62sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in {'canberra'}:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/{feature}/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/{feature}/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages[feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages[feature][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     feature][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   feature][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/{feature}/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages[feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GCV-nonredundant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = 'GCV-nonredundant'\n",
    "\n",
    "all_distances = sorted('_'.join(filename.split('_')[:-1]) \n",
    "                           for filename in os.listdir(f\"{MATRIX_DIRECTORY}/{feature}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages[feature] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     feature] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   feature] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canberra 0\n",
      "canberra 1ec\n",
      "canberra 2ec\n",
      "canberra 3ec\n",
      "canberra 4ec\n",
      "98: 15.77sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in {'canberra'}:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/{feature}/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/{feature}/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages[feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages[feature][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     feature][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   feature][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/{feature}/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages[feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GCV-orca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = 'GCV-orca'\n",
    "\n",
    "all_distances = sorted('_'.join(filename.split('_')[:-1]) \n",
    "                           for filename in os.listdir(f\"{MATRIX_DIRECTORY}/{feature}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages[feature] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     feature] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   feature] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hellinger 0\n",
      "hellinger 1c\n",
      "hellinger 2c\n",
      "hellinger 3c\n",
      "hellinger 4c\n",
      "hellinger 5c\n",
      "hellinger 6c\n",
      "hellinger 7c\n",
      "hellinger 8c\n",
      "hellinger 9c\n",
      "98: 10.39sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in {'hellinger'}:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/{feature}/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/{feature}/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages[feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages[feature][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     feature][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   feature][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/{feature}/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages[feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GCV-orca+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = 'GCV-orca+'\n",
    "\n",
    "all_distances = sorted('_'.join(filename.split('_')[:-1]) \n",
    "                           for filename in os.listdir(f\"{MATRIX_DIRECTORY}/{feature}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_coverages[feature] = defaultdict(pd.DataFrame)\n",
    "GO_coverages[     feature] = defaultdict(pd.DataFrame)\n",
    "gene_coverages[   feature] = defaultdict(pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cityblock 0\n",
      "cityblock 1c\n",
      "cityblock 2c\n",
      "cityblock 3c\n",
      "cityblock 4c\n",
      "cityblock 5c\n",
      "cityblock 6c\n",
      "cityblock 7c\n",
      "cityblock 8c\n",
      "cityblock 9c\n",
      "98: 10.41sec\n"
     ]
    }
   ],
   "source": [
    "method = 'kmedoid'\n",
    "\n",
    "for distance in {'cityblock'}:\n",
    "    \n",
    "    CLUSTER_DIRECTORY = f\"{YEAST_DIRECTORY}/clusterings/{feature}/{distance}/{method}\"\n",
    "    PVALUE_DIRECTORY = f\"{YEAST_DIRECTORY}/p-values/{feature}/{distance}/{method}/{aspect}\"\n",
    "    runs = min(get_number_of_pre_runs(PVALUE_DIRECTORY, MAX_CLUSTERS), MAX_RUNS)\n",
    "\n",
    "    for run in range(runs):\n",
    "        t1 = time.time()\n",
    "        print(f\"{distance} {run}\")\n",
    "        \n",
    "        cluster_coverages[feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2)) \n",
    "        GO_coverages[     feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        gene_coverages[   feature][distance][run] = pd.Series(np.nan, index=range(MIN_CLUSTERS, MAX_CLUSTERS+2))\n",
    "        \n",
    "        for nb_clusters in range(MIN_CLUSTERS, MAX_CLUSTERS):\n",
    "            with open(f\"{CLUSTER_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", 'r') as f:\n",
    "                 cluster_list = [set(line.split()) for line in f]\n",
    "            cluster_df = pd.Series({gene:cluster_idx \n",
    "                                        for cluster_idx,cluster in enumerate(cluster_list) \n",
    "                                        for gene in cluster})\n",
    "            \n",
    "            p_values = pd.read_csv(f\"{PVALUE_DIRECTORY}/{run}_{nb_clusters}_BioGRID.txt\", index_col=0)\n",
    "\n",
    "            enrichments = get_enrichments(alpha, p_values, cluster_list, correction)\n",
    "            enrichmet_list = [set(enrichments[i][enrichments[i]].index) for i in enrichments.columns]\n",
    "            \n",
    "            cluster_coverages[feature][distance][run][nb_clusters] = sum(enrichments.any())      / nb_clusters\n",
    "            GO_coverages[     feature][distance][run][nb_clusters] = sum(enrichments.any(axis=1))/len(GO_population)\n",
    "            gene_coverages[   feature][distance][run][nb_clusters] = sum(is_annotated_in(gene,enrichmet_list[cluster_idx])\n",
    "                                                             for gene, cluster_idx in cluster_df.items()) / len(PPI)\n",
    "            t2 = time.time()\n",
    "            print(f'{nb_clusters}: {t2-t1:.2f}sec', end='\\r')\n",
    "            \n",
    "    ENRICHMENT_DIRECTORY = f\"{YEAST_DIRECTORY}/enrichments/{feature}/{distance}/{method}/{aspect}/{correction}\"\n",
    "    if not os.path.exists(ENRICHMENT_DIRECTORY):\n",
    "        os.makedirs(ENRICHMENT_DIRECTORY)\n",
    "        \n",
    "    cluster_coverages[feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/clusters.csv\")\n",
    "    GO_coverages[     feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/GO-terms.csv\")\n",
    "    gene_coverages[   feature][distance].to_csv(f\"{ENRICHMENT_DIRECTORY}/genes.csv\")\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "graco",
   "language": "python",
   "name": "graco"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
